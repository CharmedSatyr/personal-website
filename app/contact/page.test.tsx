import { fireEvent, render, screen, within } from "@testing-library/react";
import { useRouter } from "next/navigation";

import Contact from "app/contact/page";

jest.mock("next/navigation", () => ({
	useRouter: jest.fn(),
}));

describe("Contact", () => {
	const push = jest.fn();
	const requestSubmit = jest.fn();

	beforeEach(() =>
		(useRouter as jest.Mock).mockImplementation(() => ({
			push,
			requestSubmit,
		})),
	);
	afterEach(() => jest.clearAllMocks());

	it("renders a heading", () => {
		render(<Contact />);

		expect(screen.getByRole("heading", { level: 1 })).toBeInTheDocument();
	});

	it("renders a link", () => {
		render(<Contact />);

		expect(screen.getByRole("link")).toBeInTheDocument();
	});

	it("renders a button", () => {
		render(<Contact />);

		expect(screen.getByRole("button")).toBeInTheDocument();
	});

	it("does not initially render a text box", () => {
		render(<Contact />);

		expect(screen.queryByRole("textbox")).not.toBeInTheDocument();
	});

	it("should show a challenge if the bottom link is clicked", async () => {
		render(<Contact />);

		fireEvent.click(screen.getByRole("button"));

		expect(screen.getByRole("textbox")).toBeInTheDocument();
	});

	it("should do nothing if the challenge is not completed correctly", () => {
		render(<Contact />);

		expect(screen.getAllByRole("button").length).toBe(1);

		fireEvent.click(screen.getByRole("button"));

		const input = screen.getByRole("textbox");

		fireEvent.change(input, { target: { value: "Ralph" } });

		expect(push).not.toHaveBeenCalled();
		expect(screen.queryByRole("button")).not.toBeInTheDocument();
	});

	it("should trigger a `mailto:` link if the challenge is completed correctly", () => {
		render(<Contact />);

		expect(screen.getAllByRole("button").length).toBe(1);

		fireEvent.click(screen.getByRole("button"));

		const input = screen.getByRole("textbox");

		fireEvent.change(input, { target: { value: "Joseph" } });

		expect(push).toHaveBeenCalledWith(expect.stringContaining("mailto:"));
	});

	it("should show a button that triggers a `mailto:` link if the challenge is completed correctly", async () => {
		render(<Contact />);

		fireEvent.click(screen.getByRole("button"));

		const input = screen.getByRole("textbox");

		expect(screen.queryByRole("button")).not.toBeInTheDocument();

		fireEvent.change(input, { target: { value: "Joseph" } });

		expect(push).toHaveBeenCalledTimes(1);

		const button = screen.getByRole("button");

		const emailLink = within(button).getByRole("link");

		expect(emailLink).toHaveProperty(
			"href",
			expect.stringContaining("mailto:"),
		);
	});
});
