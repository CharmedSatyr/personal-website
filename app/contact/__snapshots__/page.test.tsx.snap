// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact should not deviate unexpectedly from the last snapshot 1`] = `
<div>
  <h1
    class="title"
  >
    Contact
  </h1>
  <ol>
    <li
      class="mb-8"
    >
      <a
        class="hyperlink external"
        href="https://www.linkedin.com/in/joseph-wolfe"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          class="inline h-6 w-6 text-accent-600 dark:text-dark-accent-300"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Linkedin
          </title>
          <path
            d="M20.447 20.452h-3.554v-5.569c0-1.328-.027-3.037-1.852-3.037-1.853 0-2.136 1.445-2.136 2.939v5.667H9.351V9h3.414v1.561h.046c.477-.9 1.637-1.85 3.37-1.85 3.601 0 4.267 2.37 4.267 5.455v6.286zM5.337 7.433a2.062 2.062 0 01-2.063-2.065 2.064 2.064 0 112.063 2.065zm1.782 13.019H3.555V9h3.564v11.452zM22.225 0H1.771C.792 0 0 .774 0 1.729v20.542C0 23.227.792 24 1.771 24h20.451C23.2 24 24 23.227 24 22.271V1.729C24 .774 23.2 0 22.222 0h.003z"
          />
        </svg>
         
        Connect on LinkedIn
      </a>
    </li>
    <li
      class="mb-8"
    >
      <a
        class="hyperlink external"
        href="https://www.github.com/CharmedSatyr"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          class="inline h-6 w-6 text-accent-600 dark:text-dark-accent-300"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Github
          </title>
          <path
            d="M12 .297c-6.63 0-12 5.373-12 12 0 5.303 3.438 9.8 8.205 11.385.6.113.82-.258.82-.577 0-.285-.01-1.04-.015-2.04-3.338.724-4.042-1.61-4.042-1.61C4.422 18.07 3.633 17.7 3.633 17.7c-1.087-.744.084-.729.084-.729 1.205.084 1.838 1.236 1.838 1.236 1.07 1.835 2.809 1.305 3.495.998.108-.776.417-1.305.76-1.605-2.665-.3-5.466-1.332-5.466-5.93 0-1.31.465-2.38 1.235-3.22-.135-.303-.54-1.523.105-3.176 0 0 1.005-.322 3.3 1.23.96-.267 1.98-.399 3-.405 1.02.006 2.04.138 3 .405 2.28-1.552 3.285-1.23 3.285-1.23.645 1.653.24 2.873.12 3.176.765.84 1.23 1.91 1.23 3.22 0 4.61-2.805 5.625-5.475 5.92.42.36.81 1.096.81 2.22 0 1.606-.015 2.896-.015 3.286 0 .315.21.69.825.57C20.565 22.092 24 17.592 24 12.297c0-6.627-5.373-12-12-12"
          />
        </svg>
         
        Collaborate on GitHub
      </a>
    </li>
    <li
      class="mb-8"
    >
      <button
        class="hyperlink"
      >
        <svg
          class="inline h-6 w-6 text-accent-600 dark:text-dark-accent-300"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Mail
          </title>
          <path
            d="M2.003 5.884L10 9.882l7.997-3.998A2 2 0 0016 4H4a2 2 0 00-1.997 1.884z"
          />
          <path
            d="M18 8.118l-8 4-8-4V14a2 2 0 002 2h12a2 2 0 002-2V8.118z"
          />
        </svg>
         
        Send me an email
      </button>
    </li>
    <li>
      <a
        class="hyperlink external"
        href="https://keys.openpgp.org/vks/v1/by-fingerprint/4BFDEC92BE41B3E455B8595206DC90228603707A"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          class="inline h-6 w-6 text-accent-600 dark:text-dark-accent-300"
          fill="currentColor"
          version="1.1"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M768 341.333333h-42.666667v-85.333333c0-117.76-95.573333-213.333333-213.333333-213.333333S298.666667 138.24 298.666667 256v85.333333h-42.666667c-47.146667 0-85.333333 38.186667-85.333333 85.333334v426.666666c0 47.146667 38.186667 85.333333 85.333333 85.333334h512c47.146667 0 85.333333-38.186667 85.333333-85.333334V426.666667c0-47.146667-38.186667-85.333333-85.333333-85.333334zM512 725.333333c-47.146667 0-85.333333-38.186667-85.333333-85.333333s38.186667-85.333333 85.333333-85.333333 85.333333 38.186667 85.333333 85.333333-38.186667 85.333333-85.333333 85.333333z m132.266667-384H379.733333v-85.333333c0-72.96 59.306667-132.266667 132.266667-132.266667 72.96 0 132.266667 59.306667 132.266667 132.266667v85.333333z"
          />
        </svg>
         
        OpenPGP Key
      </a>
    </li>
  </ol>
</div>
`;
